File: 7일차/요셉/거리두기확인하기.py
# 거리두기 확인하기
# https://programmers.co.kr/learn/courses/30/lessons/81302

# 00000
# 00P12
# 0X4X0
# 00500

def check(y, x, place):
    
    # p 오른쪽이 p면 false
    if place[y][x+1] == 'P':
        return False
    
    # p 오른쪽이 o일 때, x축 p+1 하, p+2 확인
    elif place[y][x+1] == 'O':
        if place[y][x+2] == 'P' or place[y+1][x+1] == 'P':
            return False
    
    # p 하단이 p면 false
    if place[y+1][x] == 'P':
        return False
    
    # p 하단이 o일 때, y축 p+1 좌우, p+2 확인
    elif place[y+1][x] == 'O':
        if place[y+2][x] == 'P' or place[y+1][x+1] == 'P' or place[y+1][x-1] == 'P':
            return False
    
    # p 왼쪽이 o일 때, y축 p+1 좌 확인
    if place[y][x-1] == 'O' and place[y+1][x-1] == 'P':
        return False
        
    return True
        

def find(place):
    # 범위 안 벗어나도록 패딩 추가
    place = ['0{0}00'.format(p) for p in place]
    place.extend(['00000000', '00000000'])
    
    for i in range(7):
        for j in range(8):
            if place[i][j] == 'P':
                if not check(i, j, place): # 맨해튼 거리 체크
                    return False
    return True

def solution(places):
    answer = []
    
    # 순서대로
    for place in places:
        if not find(place): # 거리두기 판별
            answer.append(0)
        else:
            answer.append(1)
        
    return answer

File: 7일차/현동/거리두기확인하기.js
// https://programmers.co.kr/learn/courses/30/lessons/81302

const dy = [-1, 0, 1, 0];
const dx = [0, 1, 0, -1];

function checkDistance(y, x, level, visited, place) {
  if (place[y][x] === "X" || level === 3) return 1;
  if (level !== 0 && place[y][x] === "P") return 0;
  visited[y][x] = true;
  for (let i = 0; i < 4; i++) {
    const ny = y + dy[i];
    const nx = x + dx[i];
    if (ny < 0 || nx < 0 || ny >= 5 || nx >= 5) continue;
    if (place[ny][nx] === "X") continue;
    if (visited[ny][nx]) continue;
    if (checkDistance(ny, nx, level + 1, visited, place) === 0) return 0;
  }
}

function findP(place) {
  for (let y = 0; y < 5; y++) {
    for (let x = 0; x < 5; x++) {
      if (place[y][x] === "P") {
        let visited = Array.from({ length: 5 }, () => Array(5).fill(false));
        if (checkDistance(y, x, 0, visited, place) === 0) return 0;
      }
    }
  }
  return 1;
}

function solution(places) {
  let result = Array(5).fill(1);
  places.forEach((place, index) => {
    result[index] = findP(place);
  });
  return result;
}
